File: ruby/robot-simulator/robot_simulator.rb
class Robot

  DIRECTIONS = [:north, :east, :south, :west]

  def initialize
    at(0,0)
    orient(:north)
  end

  def at(x, y)
    @x = x
    @y = y
  end

  def orient(direction)
    raise ArgumentError, "Invalid direction" unless DIRECTIONS.include? direction
    @direction = direction
  end

  def bearing
    @direction
  end

  def coordinates
    [@x, @y]
  end

  def turn_right
    dir_idx = (DIRECTIONS.index(@direction) + 1) % DIRECTIONS.length
    orient DIRECTIONS[dir_idx]
  end

  def turn_left
    dir_idx = (DIRECTIONS.index(@direction) - 1) % DIRECTIONS.length
    orient DIRECTIONS[dir_idx]
  end

  def advance
    case @direction
    when :north then at(@x, @y+1)
    when :south then at(@x, @y-1)
    when :east then at(@x+1, @y)
    when :west then at(@x-1, @y)
    end
  end


end

class Simulator

  INSTRUCTIONS = {
    "L" => :turn_left,
    "R" => :turn_right,
    "A" => :advance
  }

  def instructions(input)
    input.chars.each_with_object([]) { |cmd, instructions| instructions << INSTRUCTIONS[cmd] }
  end

  def place(robot, options)
    robot.at(options[:x], options[:y])
    robot.orient(options[:direction])
  end

  def evaluate(robot, commands)
    instructions(commands).each { |cmd| robot.send(cmd) }
  end

end

File: elixir/robot-simulator/robot_simulator.exs
defmodule RobotSimulator do

  defstruct direction: :north, position: {0,0}

  @doc """
  Create a Robot Simulator given an initial direction and position.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec create(direction :: atom, position :: { integer, integer }) :: any
  def create(direction \\ :north, position \\ {0,0})

  def create(direction, _position) when not direction in [:north, :east, :south, :west] do
    {:error, "invalid direction"}
  end

  def create(_direction, position) when not is_tuple(position) do
    {:error, "invalid position"}
  end

  def create(_direction, position) when tuple_size(position) != 2 do
    {:error, "invalid position"}
  end

  def create(_direction, {a,b}) when not is_integer(a) or not is_integer(b) do
    {:error, "invalid position"}
  end

  def create(direction, position) do
    %RobotSimulator{direction: direction, position: position}
  end

  @doc """
  Simulate the robot's movement given a string of instructions.

  Valid instructions are: "R" (turn right), "L", (turn left), and "A" (advance)
  """
  @spec simulate(robot :: any, instructions :: String.t ) :: any
  def simulate(robot, instructions) do
    instructions
    |> String.graphemes
    |> do_simulate(robot)
  end

  defp do_simulate([], robot), do: robot

  defp do_simulate(["L"|tail], %RobotSimulator{direction: dir} = robot) do
    case dir do
      :north -> do_simulate(tail, Map.put(robot, :direction, :west))
      :west  -> do_simulate(tail, Map.put(robot, :direction, :south))
      :south -> do_simulate(tail, Map.put(robot, :direction, :east))
      :east  -> do_simulate(tail, Map.put(robot, :direction, :north))
    end
  end

  defp do_simulate(["R"|tail], %RobotSimulator{direction: dir} = robot) do
    case dir do
      :north -> do_simulate(tail, Map.put(robot, :direction, :east))
      :east  -> do_simulate(tail, Map.put(robot, :direction, :south))
      :south -> do_simulate(tail, Map.put(robot, :direction, :west))
      :west  -> do_simulate(tail, Map.put(robot, :direction, :north))
    end
  end

  defp do_simulate(["A"|tail], %RobotSimulator{direction: dir, position: {x,y}} = robot) do
    case dir do
      :north -> do_simulate(tail, Map.put(robot, :position, {x,y+1}))
      :east  -> do_simulate(tail, Map.put(robot, :position, {x+1,y}))
      :south -> do_simulate(tail, Map.put(robot, :position, {x,y-1}))
      :west  -> do_simulate(tail, Map.put(robot, :position, {x-1,y}))
    end
  end

  defp do_simulate(_instructions, _robot), do: {:error, "invalid instruction"}

  @doc """
  Return the robot's direction.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec direction(robot :: any) :: atom
  def direction(robot) do
    robot.direction
  end

  @doc """
  Return the robot's direction.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec position(robot :: any) :: { integer, integer }
  def position(robot) do
    robot.position
  end

end
